Sources/AudioKitUI/Visualizations/FFTView.swift

```diff
@@ -25,46 +26,37 @@ class FFTModel: ObservableObject {
         }
     }
 
-    func updateAmplitudes(_ fftFloats: [Float]) {
-        var fftData = fftFloats
-        for index in 0 ..< fftData.count {
-            if fftData[index].isNaN { fftData[index] = 0.0 }
-        }
+    func updateAmplitudes(_ fftData: [Float]) {
+        var tempAmplitudes = Array(repeating: 0.0 as Float, count: numberOfBars)
+        let binsPerBar = Int((Float(fftData.count) / Float(numberOfBars)).rounded(.up))
 
-        var tempAmplitudes: [Double] = []
-
-        // loop by two through all the fft data
-        for i in stride(from: 0, to: FFT_SIZE - 1, by: 2) {
-            if i / 2 < numberOfBars {
-                let binMagnitude = fftData[i]
-                let amplitude = Double(10.0 * log10(binMagnitude))
-
-                // map amplitude array to visualizer
-                var mappedAmplitude = map(n: amplitude,
-                                          start1: minAmplitude,
-                                          stop1: maxAmplitude,
-                                          start2: 0.0,
-                                          stop2: 1.0)
-                if mappedAmplitude > 1.0 {
-                    mappedAmplitude = 1.0
-                }
-                if mappedAmplitude < 0.0 {
-                    mappedAmplitude = 0.0
-                }
+        var one = Float(1.0)
+        var zero = Float(0.0)
+        var decibelNormalizationFactor = Float(1.0 / (maxAmplitude - minAmplitude))
+        var decibelNormalizationOffset = Float(-minAmplitude / (maxAmplitude - minAmplitude))
+
+        var decibels = [Float](repeating: 0, count: fftData.count)
+        vDSP_vdbcon(fftData, 1, &one, &decibels, 1, vDSP_Length(fftData.count), 0)
+        vDSP_vsmsa(
+            decibels, 1,
+            &decibelNormalizationFactor,
+            &decibelNormalizationOffset,
+            &decibels, 1,
+            vDSP_Length(decibels.count)
+        )
+        vDSP_vclip(decibels, 1, &zero, &one, &decibels, 1, vDSP_Length(decibels.count))
+
+        for (index, decibel) in decibels.enumerated() {
+            let bar = index / binsPerBar
+
+            tempAmplitudes[bar] += decibel * (1 / Float(binsPerBar))
```

Comment:
This is where we average the bins into a bar. So if there are 3 bins per bar, we'll take 1/3 of each bin. Does that make sense?

Reply:
Ahh gotcha. That does make sense.